### PR TITLE
Remove timespec

### PR DESCRIPTION
**Pull request by jralls**
This is a bit involved, so let's have a code review.

### CODE REVIEW

File: libgnucash/backend/sql/gnc-sql-column-table-entry.cpp [hunk 1]
```diff
@@ -367,151 +367,87 @@ GncSqlColumnTableEntryImpl<CT_GUID>::add_to_query(QofIdTypeConst obj_name,
...
     {
         try
         {
             auto val = row.get_string_at_col(m_col_name);
             GncDateTime time(val);
-            ts.tv_sec = static_cast<time64>(time);
+            t = static_cast<time64>(time);
         }
-        catch (const std::invalid_argument& err)
+        catch (std::invalid_argument& err)
```
Comment (gjanssens) [inline libgnucash/backend/sql/gnc-sql-column-table-entry.cpp]: Why did you remove the const qualifier ? I have been using it in my catch expressions because I saw plenty of examples using one. But honestly I haven't thought this through so I don't know which one is better or why.

File: gnucash/gnome-utils/gnc-tree-view-split-reg.c [hunk 2]
```diff
@@ -1474,96 +1476,92 @@ gtv_sr_cdf0 (GtkTreeViewColumn *col, GtkCellRenderer *cell, GtkTreeModel *s_mode
...
                 {
-                    ts.tv_sec = gnc_time (NULL);
+                    t = gnc_time (NULL);
                     //xaccSplitSetDateReconciledTS (split, ts.tv_sec);
                 }//if
-                s = gnc_print_date (ts);
+                qof_print_date_buff (datebuff, sizeof(datebuff), 0);
             }
             else
-                s = "";
```
[the remark this answers is not in the document]
Comment (jralls) [reply gnucash/gnome-utils/gnc-tree-view-split-reg.c]: You're right, the else should go too.

File: libgnucash/engine/gncBillTerm.c [hunk 3]
```diff
@@ -753,15 +753,18 @@ gboolean gncBillTermIsDirty (const GncBillTerm *term)
...
+    struct tm tm;
     int cutoff = term->cutoff;
 
     g_return_if_fail (term->type == GNC_TERM_TYPE_PROXIMO);
-
-    gnc_timespec2dmy (post_date, &iday, &imonth, &iyear);
+    gnc_localtime_r (&post_date, &tm);
+    iday = tm.tm_mday;
+    imonth = tm.tm_mon + 1;
+    iyear = tm.tm_year + 1;
```
[the remark this answers is not in the document]
Comment (jralls) [reply libgnucash/engine/gncBillTerm.c]: Right again...

File: libgnucash/engine/Transaction.c [hunk 4]
```diff
@@ -327,12 +326,10 @@ gnc_transaction_get_property(GObject* object,
         g_value_take_object(value, tx->common_currency);
         break;
     case PROP_POST_DATE:
-        ts.tv_sec = tx->date_posted;
-        g_value_set_boxed(value, &ts);
+        g_value_set_boxed(value, &tx->date_posted);
```
[the remark this answers is not in the document]
Comment (jralls) [reply libgnucash/engine/Transaction.c]: It's a C file, so it would have to be ```(Time64*) ``` rather than ```static_cast<Time64*>```, but I think if it was C++ it wouldn't even let me do ```reinterpret_cast<Time64*>```. Probably better to just write it correctly instead of playing pointer games.

File: libgnucash/engine/gnc-date.h [hunk 5]
```diff
@@ -82,6 +82,14 @@ extern "C"
...
  */
 typedef gint64 time64;
+/* A bit of a hack to create a type separate from the alias of int64_t so that
+ * compile-time dispatch can use the right KVP visitor.
+ */
+typedef struct
+{
+    time64 t;
+} Time64;
+
```
[the remark this answers is not in the document]
Comment (jralls) [reply libgnucash/engine/gnc-date.h]: The problem is that we can't differentiate between int64_t and time64 *even in C++*.  Aliases don't create new types, just new names for types, so the compiler doesn't see a difference between KvpValue<int64_t> and KvpValue<time64>.  We must use struct or class to create new types.

GValue doesn't care; in fact it depends on the sort of pointer games that you objected to in Transaction.c. But it's not feasible to separate the GValue type from the Kvp type because we pass GValues to Kvp.

[the remark this answers is not in the document]
Comment (gjanssens) [reply libgnucash/engine/gnc-date.h]: Ah, that helps. I had a vague recollection of strong typing in boost::mpl, but they do indeed use classes.

File: libgnucash/backend/sql/gnc-sql-column-table-entry.cpp (hunk 1 above)
Comment (jralls) [reply libgnucash/backend/sql/gnc-sql-column-table-entry.cpp]: I didn't exactly mean to. What I actually did was to delete the CT_TIMESPEC load and change the CT_TIME64 load, and I had to add back in the alternate string handler and missed making it const.

**Comment by jralls** [comment]
@gjanssens, all of your comments addressed and Travis now passing. On to the engine.i changes we discussed in IRC yesterday.

**Comment by jralls** [comment]
I'm having second (third?) thoughts about restoring timespec-based functionality to the python and guile bindings because a lot of the implementation requirements, in particular the timepair<->timespec  typemaps, all of the Scheme timepair code, and bindings/python/timespec.i, were removed in January. 

None of the functions here are currently bound to Python at all, and none of the Scheme bindings can be used because there's no good way to make a Timespec* to pass to any of the functions.

**Comment by gjanssens** [comment]
Ok, as none of the bindings can use these functions anyway, there's indeed no need to implement a compatibility layer for timespec.

However your PR renames many functions ending in *_t64 to the same without that extension (replacing the timespec variant) and functions with time64 in the name to only time in the name (again replacing timespec variants).

Were these *_t64 and *time64* functions exposed to the bindings? If so, I think we should keep them (but just as a wrapper of the new functions) to more or less keep the 3.x interface stable. We can try to mark them deprecated so consumers outside of the gnucash code are informed in time.

**Comment by jralls** [comment]
This latest push restores the t64/time64 function names throughout. It implements the removed timespec API in engine-deprecated.c, which is incorporated into swig-engine. There's a simple example in libgnucash/engine/test/test-extras.scm's gnc-pricedb-create function definition showing how to create a Timespec and call gnc-price-set-day with it. So far so good.

Not so good is that the calls to scm_c_issue_deprecation_warning don't seem to do anything. No warnings appear in during compilation of the tests nor in LastTest.log.

**Comment by gjanssens** [comment]
It is issuing the warning. I found it under test 77 (test-charts.scm) instead of 66 (test-extras.scm). So it appears the test logging is not fully synchronous.
By default the warning is fairly generic ("You appear to be using some deprecated guile functions", more or less). By setting environment variable GUILE_WARN_DEPRECATED=detailed I got your exact message:
gnc_price_set_time is deprecated. Use gnc_price_set_time64 instead.

So as far as I'm concerned your code is good to go in.

**Comment by gjanssens** [comment]
Eh, right... We don't need to test deprecated functions nor test whether the deprecation message works.

**Comment by christopherlam** [comment]
> It is issuing the warning. I found it under test 77 (test-charts.scm) instead of 66 (test-extras.scm). So it appears the test logging is not fully synchronous.

Correct - often my logging code and my [pass] or [fail] assertions don't match.

**Comment by gjanssens** [comment]
I agree there are no instances left in the **gnucash** code using timepair functionality. However users may have generated custom guile code (custom reports) that still use it. We don't want these users to be inconvenienced in the middle of the 3.x series. API breakage should only happen on major release increments. So we plan to drop the timepair compatibility layer in the 4.0 release.

**Comment by jralls** [comment]
Well... time*pair* got dropped in the run-up to 3.0, but much of the time*spec* API didn't, so the last commit in this set provides deprecated wrappers to time64 for that remaining timespec API.
